crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -2469,7 +2477,24 @@ impl<'db> TypeInferenceBuilder<'db> {
             (UnaryOp::Invert, Type::BooleanLiteral(bool)) => Type::IntLiteral(!i64::from(bool)),
 
             (UnaryOp::Not, ty) => ty.bool(self.db).negate().into_type(self.db),
+            (_, Type::Any) => Type::Any,
+            (_, Type::Unknown) => Type::Unknown,
+            (op @ (UnaryOp::UAdd | UnaryOp::USub | UnaryOp::Invert), Type::Instance(class)) => {
+                let class_member = class.class_member(self.db, op.dunder());
+                let call = class_member.call(self.db, &[operand_type]);
+                if matches!(call, CallOutcome::NotCallable { not_callable_ty: _ }) {
+                    self.add_diagnostic(
+                        unary.into(),
+                        "unsupported-operator",
+                        format_args!(
+                            "Operator `{op}` is unsupported for type `{}`",
```

Comment:
Minor nit: since `+` and `-` are both unary and binary operators, let's make sure this diagnostic is clear which one we're talking about:
```suggestion
                            "Unary operator `{op}` is unsupported for type `{}`",
```